cdap-app-fabric/src/main/java/co/cask/cdap/metadata/LineageAdmin.java

```diff
@@ -217,8 +259,80 @@ private Lineage doComputeLineage(final NamespacedEntityId sourceData, long start
+      Map<ProgramRunId, RunRecordMeta> workflowRunRecordMap =
+        store.getRuns(ProgramRunStatus.ALL,
+                      new Predicate<RunRecordMeta>() {
+                        @Override
+                        public boolean apply(RunRecordMeta input) {
+                          return workflowIDs.contains(input.getPid());
+                        }
+                      });
+
+      // Create a map from RunId to ProgramId for all workflows
+      Map<String, ProgramRunId> workflowIdMap = new HashMap<>();
+      for (Map.Entry<ProgramRunId, RunRecordMeta> entry : workflowRunRecordMap.entrySet()) {
+        workflowIdMap.put(entry.getValue().getPid(), entry.getKey());
+      }
+
+      // For all relations, replace ProgramIds with workflow ProgramIds
+      Multimap<RelationKey, Relation> relationsNew = HashMultimap.create();
+
+      for (RelationKey key : relations.keySet()) {
+        Collection<Relation> allRelations = relations.get(key);
+        for (Relation relation : allRelations) {
+          ProgramId program = relation.getProgram();
+          RunRecordMeta runRecord = store.getRun(program, relation.getRun().getId());
+          if (runRecord != null
+            && runRecord.getProperties().containsKey("workflowrunid")) {
+            String workflowRunId = runRecord.getProperties().get("workflowrunid");
+            NamespacedEntityId data = relation.getData();
+            ProgramRunId workflowProgramRunId = workflowIdMap.get(workflowRunId);
+            if (workflowProgramRunId != null) {
+              ProgramId workflowProgramId = new ProgramId(workflowProgramRunId.getNamespace(),
+                                                          workflowProgramRunId.getApplication(),
+                                                          workflowProgramRunId.getType(),
+                                                          workflowProgramRunId.getProgram());
+              Relation workflowRelation;
+              if (data instanceof DatasetId) {
+                workflowRelation = new Relation((DatasetId) data, workflowProgramId,
+                                                relation.getAccess(), RunIds.fromString(workflowRunId));
+              } else {
+                workflowRelation = new Relation((StreamId) data, workflowProgramId,
+                                                relation.getAccess(), RunIds.fromString(workflowRunId));
+              }
+              relationsNew.put(key, workflowRelation);
+            } else {
+              relationsNew.put(key, relation);
+            }
+          } else {
+            relationsNew.put(key, relation);
+          }
+        }
+      }
+      lineage = new Lineage(
+        Iterables.concat(Maps.transformValues(relationsNew.asMap(), COLLAPSE_UNKNOWN_TYPE_FUNCTION).values()));
+
+    } else {
+      lineage = new Lineage(
+        Iterables.concat(Maps.transformValues(relations.asMap(), COLLAPSE_UNKNOWN_TYPE_FUNCTION).values()));
```

Comment:
Since the `else` case is so small, the flow would be easier to read and understand if you flip the condition.

```java
if (!rollup.contains("workflow")) {
  lineage = new Lineage(...);
  LOG.trace(...);
  return lineage;
}

// The rollup by workflow logic here.
final Set<String> workflowIDs = new HashSet<>();
...
```